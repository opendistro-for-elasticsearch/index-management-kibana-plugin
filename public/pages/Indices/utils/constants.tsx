/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from "react";
import { EuiHealth, EuiTableFieldDataColumnType } from "@elastic/eui";
import { ManagedCatIndex } from "../../../../server/models/interfaces";
import { SortDirection } from "../../../utils/constants";

export const DEFAULT_PAGE_SIZE_OPTIONS = [5, 10, 20, 50];

export const DEFAULT_QUERY_PARAMS = {
  from: 0,
  size: 20,
  search: "",
  sortField: "name",
  sortDirection: SortDirection.DESC,
};

const HEALTH_TO_COLOR: {
  [health: string]: string;
  green: string;
  yellow: string;
  red: string;
} = {
  green: "success",
  yellow: "warning",
  red: "danger",
};

export const indicesColumns: EuiTableFieldDataColumnType<ManagedCatIndex>[] = [
  {
    field: "index",
    name: "Index",
    sortable: true,
    truncateText: true,
    textOnly: true,
    width: "250px",
    render: (index: string) => <span title={index}>{index}</span>,
  },
  {
    field: "health",
    name: "Health",
    sortable: true,
    truncateText: true,
    textOnly: true,
    align: "right",
    render: (health: string, item: ManagedCatIndex) => {
      const color = health ? HEALTH_TO_COLOR[health] : "subdued";
      const text = health || item.status;
      return <EuiHealth color={color}>{text}</EuiHealth>;
    },
  },
  {
    field: "managed",
    name: "Managed by Policy",
    sortable: false,
    truncateText: true,
    textOnly: true,
    align: "right",
    width: "150px",
  },
  {
    field: "status",
    name: "Status",
    sortable: true,
    truncateText: true,
    textOnly: true,
    align: "right",
  },
  {
    field: "store.size",
    name: "Total size",
    sortable: true,
    truncateText: true,
    textOnly: true,
    dataType: "number",
  },
  {
    field: "pri.store.size",
    name: "Primaries size",
    sortable: true,
    truncateText: true,
    textOnly: true,
    dataType: "number",
  },
  {
    field: "docs.count",
    name: "Total documents",
    sortable: true,
    truncateText: true,
    textOnly: true,
    dataType: "number",
    render: (count: string) => <span title={count}>{count}</span>,
  },
  {
    field: "docs.deleted",
    name: "Deleted documents",
    sortable: true,
    truncateText: true,
    textOnly: true,
    dataType: "number",
    render: (deleted: string) => <span title={deleted}>{deleted}</span>,
  },
  {
    field: "pri",
    name: "Primaries",
    sortable: true,
    truncateText: true,
    textOnly: true,
    dataType: "number",
  },
  {
    field: "rep",
    name: "Replicas",
    sortable: true,
    truncateText: true,
    textOnly: true,
    dataType: "number",
  },
];
