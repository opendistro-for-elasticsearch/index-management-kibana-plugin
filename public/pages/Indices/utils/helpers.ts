/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import queryString from "query-string";
import { DEFAULT_QUERY_PARAMS } from "./constants";
import { IndicesQueryParams } from "../models/interfaces";

export function getURLQueryParams(location: { search: string }): IndicesQueryParams {
  const { from, size, search, sortField, sortDirection } = queryString.parse(location.search);
  return <IndicesQueryParams>{
    // @ts-ignore
    from: isNaN(parseInt(from, 10)) ? DEFAULT_QUERY_PARAMS.from : parseInt(from, 10),
    // @ts-ignore
    size: isNaN(parseInt(size, 10)) ? DEFAULT_QUERY_PARAMS.size : parseInt(size, 10),
    search: typeof search !== "string" ? DEFAULT_QUERY_PARAMS.search : search,
    sortField: typeof sortField !== "string" ? "index" : sortField,
    sortDirection: typeof sortDirection !== "string" ? DEFAULT_QUERY_PARAMS.sortDirection : sortDirection,
  };
}
