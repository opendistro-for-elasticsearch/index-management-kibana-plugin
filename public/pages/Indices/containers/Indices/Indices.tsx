/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { Component } from "react";
import _ from "lodash";
import { RouteComponentProps } from "react-router-dom";
import queryString from "querystring";
import {
  EuiBasicTable,
  EuiHorizontalRule,
  // @ts-ignore
  Criteria,
  EuiTableSortingType,
  Direction,
  // @ts-ignore
  Pagination,
  EuiTableSelectionType,
} from "@elastic/eui";
import { CoreStart } from "kibana/public";
import { ContentPanel, ContentPanelActions } from "../../../../components/ContentPanel";
import IndexControls from "../../components/IndexControls";
import ApplyPolicyModal from "../../components/ApplyPolicyModal";
import IndexEmptyPrompt from "../../components/IndexEmptyPrompt";
import { DEFAULT_PAGE_SIZE_OPTIONS, DEFAULT_QUERY_PARAMS, indicesColumns } from "../../utils/constants";
import { ModalConsumer } from "../../../../components/Modal";
import IndexService from "../../../../services/IndexService";
import { ManagedCatIndex } from "../../../../../server/models/interfaces";
import { getURLQueryParams } from "../../utils/helpers";
import { IndicesQueryParams } from "../../models/interfaces";
import { BREADCRUMBS } from "../../../../utils/constants";
import { getErrorMessage } from "../../../../utils/helpers";

interface IndicesProps extends RouteComponentProps {
  indexService: IndexService;
  core: CoreStart;
}

interface IndicesState {
  totalIndices: number;
  from: number;
  size: number;
  search: string;
  sortField: keyof ManagedCatIndex;
  sortDirection: Direction;
  selectedItems: ManagedCatIndex[];
  indices: ManagedCatIndex[];
  loadingIndices: boolean;
}

export default class Indices extends Component<IndicesProps, IndicesState> {
  constructor(props: IndicesProps) {
    super(props);

    const { from, size, search, sortField, sortDirection } = getURLQueryParams(this.props.location);

    this.state = {
      totalIndices: 0,
      from,
      size,
      search,
      sortField,
      sortDirection,
      selectedItems: [],
      indices: [],
      loadingIndices: true,
    };

    this.getIndices = _.debounce(this.getIndices, 500, { leading: true });
  }

  async componentDidMount() {
    // chrome.breadcrumbs.set([BREADCRUMBS.INDEX_MANAGEMENT, BREADCRUMBS.INDICES]);
    this.props.core.chrome.setBreadcrumbs([BREADCRUMBS.INDEX_MANAGEMENT, BREADCRUMBS.INDICES]);
    await this.getIndices();
  }

  async componentDidUpdate(prevProps: IndicesProps, prevState: IndicesState) {
    const prevQuery = Indices.getQueryObjectFromState(prevState);
    const currQuery = Indices.getQueryObjectFromState(this.state);
    if (!_.isEqual(prevQuery, currQuery)) {
      await this.getIndices();
    }
  }

  static getQueryObjectFromState({ from, size, search, sortField, sortDirection }: IndicesState): IndicesQueryParams {
    return { from, size, search, sortField, sortDirection };
  }

  getIndices = async (): Promise<void> => {
    this.setState({ loadingIndices: true });
    try {
      const { indexService, history } = this.props;
      const queryObject = Indices.getQueryObjectFromState(this.state);
      const queryParamsString = queryString.stringify(queryObject);
      history.replace({ ...this.props.location, search: queryParamsString });
      const getIndicesResponse = await indexService.getIndices(queryObject);
      if (getIndicesResponse.ok) {
        const { indices, totalIndices } = getIndicesResponse.response;
        this.setState({ indices, totalIndices });
      } else {
        this.props.core.notifications.toasts.addDanger(getIndicesResponse.error);
      }
    } catch (err) {
      this.props.core.notifications.toasts.addDanger(getErrorMessage(err, "There was a problem loading the indices"));
    }
    this.setState({ loadingIndices: false });
  };

  onTableChange = ({ page: tablePage, sort }: Criteria<ManagedCatIndex>): void => {
    const { index: page, size } = tablePage;
    const { field: sortField, direction: sortDirection } = sort;
    this.setState({ from: page * size, size, sortField, sortDirection });
  };

  onSelectionChange = (selectedItems: ManagedCatIndex[]): void => {
    this.setState({ selectedItems });
  };

  onSearchChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    this.setState({ from: 0, search: e.target.value });
  };

  onPageClick = (page: number): void => {
    const { size } = this.state;
    this.setState({ from: page * size });
  };

  resetFilters = (): void => {
    this.setState({ search: DEFAULT_QUERY_PARAMS.search });
  };

  render() {
    const { totalIndices, from, size, search, sortField, sortDirection, selectedItems, indices, loadingIndices } = this.state;

    const filterIsApplied = !!search;
    const page = Math.floor(from / size);

    const pagination: Pagination = {
      pageIndex: page,
      pageSize: size,
      pageSizeOptions: DEFAULT_PAGE_SIZE_OPTIONS,
      totalItemCount: totalIndices,
    };

    const sorting: EuiTableSortingType<ManagedCatIndex> = {
      sort: {
        direction: sortDirection,
        field: sortField,
      },
    };

    const selection: EuiTableSelectionType<ManagedCatIndex> = {
      onSelectionChange: this.onSelectionChange,
    };
    return (
      <ContentPanel
        actions={
          <ModalConsumer>
            {({ onShow }) => (
              <ContentPanelActions
                actions={[
                  {
                    text: "Apply policy",
                    buttonProps: {
                      disabled: !selectedItems.length,
                      onClick: () => onShow(ApplyPolicyModal, { indices: selectedItems.map((item: ManagedCatIndex) => item.index) }),
                    },
                  },
                ]}
              />
            )}
          </ModalConsumer>
        }
        bodyStyles={{ padding: "initial" }}
        title="Indices"
      >
        <IndexControls
          activePage={page}
          pageCount={Math.ceil(totalIndices / size) || 1}
          search={search}
          onSearchChange={this.onSearchChange}
          onPageClick={this.onPageClick}
          onRefresh={this.getIndices}
        />

        <EuiHorizontalRule margin="xs" />

        <EuiBasicTable
          columns={indicesColumns}
          isSelectable={true}
          itemId="index"
          items={indices}
          noItemsMessage={<IndexEmptyPrompt filterIsApplied={filterIsApplied} loading={loadingIndices} resetFilters={this.resetFilters} />}
          onChange={this.onTableChange}
          pagination={pagination}
          selection={selection}
          sorting={sorting}
        />
      </ContentPanel>
    );
  }
}
