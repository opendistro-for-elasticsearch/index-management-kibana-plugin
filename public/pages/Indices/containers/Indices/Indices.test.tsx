/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, fireEvent, wait } from "@testing-library/react";
// @ts-ignore
import userEvent from "@testing-library/user-event";
import { Redirect, Route, Switch } from "react-router";
import { HashRouter as Router } from "react-router-dom";
import { toastNotifications } from "ui/notify";
import chrome from "ui/chrome";
import { browserServicesMock } from "../../../../../test/mocks";
import Indices from "./Indices";
import { TEXT } from "../../components/IndexEmptyPrompt/IndexEmptyPrompt";
import { ModalProvider, ModalRoot } from "../../../../components/Modal";
import { ServicesConsumer, ServicesContext } from "../../../../services";
import { BREADCRUMBS, ROUTES } from "../../../../utils/constants";

function renderWithRouter(Component: React.ComponentType<any>) {
  return {
    ...render(
      <Router>
        <Switch>
          <Route
            path={ROUTES.INDICES}
            render={props => (
              <ServicesContext.Provider value={browserServicesMock}>
                <ModalProvider>
                  <ServicesConsumer>{services => services && <ModalRoot services={services} />}</ServicesConsumer>
                  <ServicesConsumer>{({ indexService }: any) => <Component indexService={indexService} {...props} />}</ServicesConsumer>
                </ModalProvider>
              </ServicesContext.Provider>
            )}
          />
          <Redirect from="/" to={ROUTES.INDICES} />
        </Switch>
      </Router>
    ),
  };
}

describe("<Indices /> spec", () => {
  it("renders the component", async () => {
    browserServicesMock.indexService.getIndices = jest.fn().mockResolvedValue({ ok: true, response: { indices: [], totalIndices: 0 } });
    const { container } = renderWithRouter(Indices);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("shows LOADING on mount", async () => {
    browserServicesMock.indexService.getIndices = jest.fn().mockResolvedValue({ ok: true, response: { indices: [], totalIndices: 0 } });
    const { getByText } = renderWithRouter(Indices);

    getByText(TEXT.LOADING);
  });

  it("sets breadcrumbs when mounting", async () => {
    browserServicesMock.indexService.getIndices = jest.fn().mockResolvedValue({ ok: true, response: { indices: [], totalIndices: 0 } });
    renderWithRouter(Indices);

    expect(chrome.breadcrumbs.set).toHaveBeenCalledTimes(1);
    expect(chrome.breadcrumbs.set).toHaveBeenCalledWith([BREADCRUMBS.INDEX_MANAGEMENT, BREADCRUMBS.INDICES]);
  });

  it("loads indices", async () => {
    const indices = [
      {
        "docs.count": 5,
        "docs.deleted": 2,
        health: "green",
        index: "index_1",
        pri: "1",
        "pri.store.size": "100KB",
        rep: "0",
        status: "open",
        "store.size": "100KB",
        uuid: "some_uuid",
      },
    ];
    browserServicesMock.indexService.getIndices = jest.fn().mockResolvedValue({ ok: true, response: { indices, totalIndices: 1 } });
    const { getByText } = renderWithRouter(Indices);
    await wait();

    await wait(() => getByText("index_1"));
  });

  it("adds error toaster when get indices has error", async () => {
    browserServicesMock.indexService.getIndices = jest.fn().mockResolvedValue({ ok: false, error: "some error" });
    renderWithRouter(Indices);

    await wait();

    expect(toastNotifications.addDanger).toHaveBeenCalledTimes(1);
    expect(toastNotifications.addDanger).toHaveBeenCalledWith("some error");
  });

  it("adds error toaster when get indices throws error", async () => {
    browserServicesMock.indexService.getIndices = jest.fn().mockRejectedValue(new Error("rejected error"));
    renderWithRouter(Indices);

    await wait();

    expect(toastNotifications.addDanger).toHaveBeenCalledTimes(1);
    expect(toastNotifications.addDanger).toHaveBeenCalledWith("rejected error");
  });

  it("can add a policy to an index", async () => {
    const indices = [
      {
        "docs.count": 5,
        "docs.deleted": 2,
        health: "green",
        index: "index_1",
        pri: "1",
        "pri.store.size": "100KB",
        rep: "0",
        status: "open",
        "store.size": "100KB",
        uuid: "some_uuid",
      },
    ];
    browserServicesMock.indexService.getIndices = jest.fn().mockResolvedValue({ ok: true, response: { indices, totalIndices: 1 } });
    browserServicesMock.indexService.searchPolicies = jest.fn().mockResolvedValue({ ok: true, response: { policies: ["some_policy"] } });
    const { getByText, getByTestId } = renderWithRouter(Indices);

    await wait(() => getByText("index_1"));

    expect(getByTestId("Apply policyButton")).toBeDisabled();

    userEvent.click(getByTestId("checkboxSelectRow-index_1"));

    expect(getByTestId("Apply policyButton")).toBeEnabled();

    userEvent.click(getByTestId("Apply policyButton"));

    await wait();

    /*
     * TODO: Cannot proceed with this test, throws error
     *   Need to figure out why combo box is complaining
     *   TypeError: Cannot read property 'position' of null
     *   at Object.updatePosition (node_modules/@elastic/eui/lib/components/combo_box/combo_box.js:123:43)
     *   at node_modules/@elastic/eui/lib/components/combo_box/combo_box_options_list/combo_box_options_list.js:106:21
     *   at Timeout.callback [as _onTimeout] (../../kibana/node_modules/jsdom/lib/jsdom/browser/Window.js:592:19)
     * */

    // userEvent.click(getByTestId("comboBoxToggleListButton"));

    // await wait();

    // fireEvent.keyPress(getByTestId("comboBoxSearchInput"), { key: "Down Arrow", code: 40, charCode: 40 });

    // await wait();
  });

  it("sorts/paginates the table", async () => {
    const indices = new Array(40).fill(null).map((_, index) => ({
      "docs.count": index,
      "docs.deleted": 2,
      health: "green",
      index: `index_${index}`,
      pri: "1",
      "pri.store.size": "100KB",
      rep: "0",
      status: "open",
      "store.size": "100KB",
      uuid: "some_uuid",
    }));
    browserServicesMock.indexService.getIndices = jest
      .fn()
      .mockResolvedValueOnce({ ok: true, response: { indices: indices.slice(0, 20), totalIndices: 40 } })
      .mockResolvedValueOnce({ ok: true, response: { indices: indices.slice(20), totalIndices: 40 } })
      .mockResolvedValueOnce({
        ok: true,
        response: {
          indices: indices.sort((a, b) => a["docs.count"] - b["docs.count"]).slice(0, 20),
          totalIndices: 40,
        },
      });

    browserServicesMock.indexService.searchPolicies = jest.fn().mockResolvedValue({ ok: true, response: { policies: ["some_policy"] } });
    const { getByText, getByTestId, getAllByTestId, queryByText } = renderWithRouter(Indices);

    // should load indices 0-19 on first load
    await wait(() => getByText("index_0"));
    expect(queryByText("index_39")).toBeNull();

    fireEvent.click(getAllByTestId("pagination-button-next")[0]);

    // should load indices 20-39 after clicking next
    await wait(() => getByText("index_39"));
    expect(queryByText("index_0")).toBeNull();

    // @ts-ignore
    fireEvent.click(getByTestId("tableHeaderCell_docs.count_6").firstChild);

    // should load indices 0-19 after clicking sort (defaults to asc) on docs.count
    await wait(() => getByText("index_0"));
    expect(queryByText("index_39")).toBeNull();
  });
});
