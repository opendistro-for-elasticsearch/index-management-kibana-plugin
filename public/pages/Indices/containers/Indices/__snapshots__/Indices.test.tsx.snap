// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Indices /> spec renders the component 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium"
  style="padding-left: 0px; padding-right: 0px;"
>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
    style="padding: 0px 10px;"
  >
    <div
      class="euiFlexItem"
    >
      <h3
        class="euiTitle euiTitle--large"
      >
        Indices
      </h3>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <button
                class="euiButton euiButton--primary euiButton-isDisabled"
                data-test-subj="Apply policyButton"
                disabled=""
                type="button"
              >
                <span
                  class="euiButtonContent euiButton__content"
                >
                  <span
                    class="euiButton__text"
                  >
                    Apply policy
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr
    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
  />
  <div
    style="padding: initial;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px 5px;"
    >
      <div
        class="euiFlexItem"
      >
        <div
          class="euiFormControlLayout euiFormControlLayout--fullWidth"
        >
          <div
            class="euiFormControlLayout__childrenWrapper"
          >
            <input
              class="euiFieldSearch euiFieldSearch--fullWidth euiFieldSearch-isClearable"
              placeholder="Search"
              type="search"
              value=""
            />
            <div
              class="euiFormControlLayoutIcons"
            >
              <span
                class="euiFormControlLayoutCustomIcon"
              >
                EuiIconMock
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
        style="max-width: 250px;"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--directionRow euiSuperDatePicker__flexWrapper euiSuperDatePicker__flexWrapper--isAutoRefreshOnly"
        >
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormControlLayout euiFormControlLayout--group euiSuperDatePicker"
            >
              <div
                class="euiPopover euiPopover--anchorDownLeft"
                id="QuickSelectPopover"
              >
                <div
                  class="euiPopover__anchor euiQuickSelectPopover__anchor"
                >
                  <button
                    aria-label="Date quick select"
                    class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiFormControlLayout__prepend"
                    data-test-subj="superDatePickerToggleQuickMenuButton"
                    type="button"
                  >
                    <span
                      class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                    >
                      EuiIconMock
                      <span
                        class="euiButtonEmpty__text euiQuickSelectPopover__buttonText"
                      >
                        EuiIconMock
                      </span>
                    </span>
                  </button>
                </div>
              </div>
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <div
                  class="euiDatePickerRange euiDatePickerRange--readOnly euiDatePickerRange--inGroup"
                >
                  <span
                    class="euiSuperDatePicker__prettyFormat"
                  >
                    Off
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      class="euiBasicTable"
    >
      <div>
        <div
          class="euiTableHeaderMobile"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <div
                class="euiCheckbox"
              >
                <input
                  aria-label="Select all rows"
                  class="euiCheckbox__input"
                  disabled=""
                  id="_selection_column-checkbox_some_html_id"
                  type="checkbox"
                />
                <div
                  class="euiCheckbox__square"
                />
                <label
                  class="euiCheckbox__label"
                  for="_selection_column-checkbox_some_html_id"
                >
                  Select all rows
                </label>
              </div>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <div
                class="euiTableSortMobile"
              >
                <div
                  class="euiPopover euiPopover--anchorDownRight"
                >
                  <div
                    class="euiPopover__anchor"
                  >
                    <button
                      class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiButtonEmpty--flushRight"
                      type="button"
                    >
                      <span
                        class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                      >
                        EuiIconMock
                        <span
                          class="euiButtonEmpty__text"
                        >
                          Sorting
                        </span>
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <table
          class="euiTable euiTable--responsive"
          id="some_html_id"
          tabindex="-1"
        >
          <caption
            class="euiScreenReaderOnly euiTableCaption"
          />
          <thead>
            <tr>
              <th
                class="euiTableHeaderCellCheckbox"
                scope="col"
              >
                <div
                  class="euiTableCellContent"
                >
                  <div
                    class="euiCheckbox euiCheckbox--inList euiCheckbox--noLabel"
                  >
                    <input
                      aria-label="Select all rows"
                      class="euiCheckbox__input"
                      data-test-subj="checkboxSelectAll"
                      disabled=""
                      id="_selection_column-checkbox_some_html_id"
                      type="checkbox"
                    />
                    <div
                      class="euiCheckbox__square"
                    />
                  </div>
                </div>
              </th>
              <th
                aria-live="polite"
                aria-sort="descending"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_index_0"
                role="columnheader"
                scope="col"
                style="width: 250px;"
              >
                <button
                  class="euiTableHeaderButton euiTableHeaderButton-isSorted"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Index; Sorted in descending order"
                    >
                      Index
                    </span>
                    EuiIconMock
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in ascending order
                    </span>
                  </span>
                </button>
              </th>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_health_1"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent euiTableCellContent--alignRight"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Health"
                    >
                      Health
                    </span>
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in ascending order
                    </span>
                  </span>
                </button>
              </th>
              <th
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_managed_2"
                role="columnheader"
                scope="col"
                style="width: 150px;"
              >
                <div
                  class="euiTableCellContent euiTableCellContent--alignRight"
                >
                  <span
                    class="euiTableCellContent__text"
                    title="Managed by Policy"
                  >
                    Managed by Policy
                  </span>
                </div>
              </th>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_status_3"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent euiTableCellContent--alignRight"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Status"
                    >
                      Status
                    </span>
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in ascending order
                    </span>
                  </span>
                </button>
              </th>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_store.size_4"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent euiTableCellContent--alignRight"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Total size"
                    >
                      Total size
                    </span>
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in ascending order
                    </span>
                  </span>
                </button>
              </th>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_pri.store.size_5"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent euiTableCellContent--alignRight"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Primaries size"
                    >
                      Primaries size
                    </span>
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in ascending order
                    </span>
                  </span>
                </button>
              </th>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_docs.count_6"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent euiTableCellContent--alignRight"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Total documents"
                    >
                      Total documents
                    </span>
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in ascending order
                    </span>
                  </span>
                </button>
              </th>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_docs.deleted_7"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent euiTableCellContent--alignRight"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Deleted documents"
                    >
                      Deleted documents
                    </span>
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in ascending order
                    </span>
                  </span>
                </button>
              </th>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_pri_8"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent euiTableCellContent--alignRight"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Primaries"
                    >
                      Primaries
                    </span>
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in ascending order
                    </span>
                  </span>
                </button>
              </th>
              <th
                aria-live="polite"
                aria-sort="none"
                class="euiTableHeaderCell"
                data-test-subj="tableHeaderCell_rep_9"
                role="columnheader"
                scope="col"
              >
                <button
                  class="euiTableHeaderButton"
                  data-test-subj="tableHeaderSortButton"
                  type="button"
                >
                  <span
                    class="euiTableCellContent euiTableCellContent--alignRight"
                  >
                    <span
                      class="euiTableCellContent__text"
                      title="Replicas"
                    >
                      Replicas
                    </span>
                    <span
                      class="euiScreenReaderOnly"
                    >
                      Click to sort in ascending order
                    </span>
                  </span>
                </button>
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="euiTableRow"
            >
              <td
                class="euiTableRowCell euiTableRowCell--isMobileFullWidth"
                colspan="11"
              >
                <div
                  class="euiTableCellContent euiTableCellContent--alignCenter"
                >
                  <span
                    class="euiTableCellContent__text"
                  >
                    <div
                      class="euiEmptyPrompt"
                      style="max-width: 45em;"
                    >
                      <span
                        class="euiTextColor euiTextColor--subdued"
                      >
                        <div
                          class="euiText euiText--medium"
                        >
                          <div
                            class="euiText euiText--medium"
                          >
                            <p>
                              Loading indices...
                            </p>
                          </div>
                        </div>
                      </span>
                    </div>
                  </span>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
